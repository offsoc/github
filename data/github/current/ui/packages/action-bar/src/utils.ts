import type {Density} from './types'

const columnGap = (el: Element | null) => {
  if (!el) return 0
  const computedStyle = window.getComputedStyle(el)
  if (computedStyle?.columnGap) return parseInt(computedStyle.columnGap, 10) // e.g., '16px' => 16
  return 0
}

export const calculateItemOffsetWidth = (item: Element, itemContainer: HTMLDivElement | null) => {
  const gap = columnGap(itemContainer) * 2
  const width = item.getBoundingClientRect().width
  const itemStyle = window.getComputedStyle(item)
  const marginLeft = itemStyle?.marginLeft ? parseInt(itemStyle.marginLeft, 10) : 0
  const marginRight = itemStyle?.marginRight ? parseInt(itemStyle.marginRight, 10) : 0
  return width + marginLeft + marginRight + gap
}

/**
 * How much space is the overflow menu toggle button taking up?
 */
export const calculateMenuSpace = (outerContainer: HTMLDivElement | null, itemContainer: HTMLDivElement | null) => {
  const availableSpace = calculateAvailableSpace(outerContainer, itemContainer)
  if (typeof availableSpace === 'undefined') return
  let menuSpace = availableSpace
  if (itemContainer) menuSpace -= itemContainer.offsetLeft
  return menuSpace
}

/**
 * How much space do we have to expand into, to show more action items?
 * @returns Get the offset of the item container from the container edge
 */
export const calculateAvailableSpace = (
  outerContainer: HTMLDivElement | null,
  itemContainer: HTMLDivElement | null,
) => {
  if (!outerContainer) return
  let availableSpace = outerContainer.offsetWidth
  if (itemContainer) availableSpace -= itemContainer.offsetWidth
  return availableSpace
}

export const gapFromDensity = (density?: Density) => {
  if (density === 'condensed') return 1
  if (density === 'spacious') return 3
  if (density === 'none') return 0
  return 2
}
