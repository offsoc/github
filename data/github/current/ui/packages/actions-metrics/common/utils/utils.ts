import {COLUMNS, TIME_IN_MS} from '../constants/constants'
import type {
  MetricsResponseItem,
  MetricsItem,
  MetricsResponse,
  MetricsView,
  Timestamp,
  CardHeading,
} from '../models/models'
import {LABELS} from '../resources/labels'
import {ScopeType} from '../models/enums'

export class Utils {
  public static objectEquals = (a?: object, b?: object, propsToBeChecked?: string[]): boolean => {
    if (!propsToBeChecked?.length) {
      // may want to use lodash deep isEqual or something similar later since stringify may be inconsistent
      return JSON.stringify(a) === JSON.stringify(b)
    }
    if (!a || !b) {
      return a === b
    }

    const aTrimmed = Utils.withOnlyRequestedKeys<object>(a, propsToBeChecked)
    const bTrimmed = Utils.withOnlyRequestedKeys<object>(b, propsToBeChecked)

    return this.objectEquals(aTrimmed, bTrimmed)
  }

  public static getCurrentPage = (offset: number, pageSize: number, totalCount?: number): number => {
    const nextItemOffset = offset
    const currentPage = Math.ceil(nextItemOffset / pageSize)
    const lastPage = totalCount ? Math.floor(totalCount / pageSize) : currentPage
    return currentPage > lastPage ? lastPage : currentPage
  }

  public static getOffsetFromPage = (pageSize: number, page: number): number => {
    return (page - 1) * pageSize + 1
  }

  public static getRealOffsetFromVirtual = (virtualOffset: number, realPageSize: number): number => {
    return Math.floor(virtualOffset / realPageSize) * realPageSize
  }

  public static convertFromResponse = <From extends MetricsResponseItem, To extends MetricsItem>(
    response: MetricsResponse<From>,
    transformer?: (from: From, uniqueId: string) => To,
  ): [MetricsView, To[]] => {
    const view = {} as MetricsView

    for (const key of Object.keys(response)) {
      if (key !== 'items') {
        Utils.convertFromResponseProperty(response, view, key)
      }
    }

    return [
      view,
      response.items.map((item, index) =>
        transformer ? transformer(item, index.toString()) : Utils.convertFromResponseItem(item, index.toString()),
      ),
    ]
  }

  public static convertFromResponseItem = <From extends MetricsResponseItem, To extends MetricsItem>(
    from: From,
    uniqueId: string,
  ): To => {
    const item = {} as To

    for (const key of Object.keys(from)) {
      Utils.convertFromResponseProperty(from, item, key)
    }

    item.id = uniqueId
    return item
  }

  public static convertFromResponseProperty = <
    From extends MetricsResponseItem | MetricsResponse<MetricsResponseItem>,
    To,
  >(
    from: From,
    to: To,
    propKey: string,
  ): void => {
    let fromProp = from[propKey as keyof From] as never | undefined | null

    if (fromProp === null) {
      fromProp = undefined
    }

    const key = Utils.snakeToCamel(propKey)

    if (key === 'startTime' || key === 'endTime') {
      const time = (fromProp as Timestamp | undefined)?.seconds
        ? (fromProp as Timestamp | undefined)!.seconds * 1000 // multiplied by 1000 so that the argument is in milliseconds, not seconds
        : Date.now() // use now as fallback in case of empty response (error)

      to[key as keyof To] = new Date(time) as never
    } else {
      to[key as keyof To] = fromProp as never
    }
  }

  public static snakeToCamel = (snake: string): string => {
    return snake
      .split('_')
      .map((part, index) => (part.length > 1 && index > 0 ? part.charAt(0).toUpperCase() + part.slice(1) : part))
      .join('')
  }

  public static camelToSnake = (camel: string): string => {
    let snake = ''

    for (let i = 0; i < camel.length; i++) {
      const c = camel.charAt(i)

      if (c === c.toUpperCase()) {
        snake += `_${c.toLowerCase()}`
      } else {
        snake += c
      }
    }

    return snake
  }

  public static camelToSnakeObject = (obj: object): object => {
    const result = {}

    for (const key of Object.keys(obj)) {
      result[Utils.camelToSnake(key) as keyof object] = obj[key as keyof object]
    }

    return result
  }

  public static in15Minutes = (minutesOverride: number = 15): Date => {
    return new Date(minutesOverride * 60000 + Date.now())
  }

  public static removeKeys = (obj: object, keys: string[]) => {
    for (const key of keys) {
      if (Object.hasOwn(obj, key)) {
        delete obj[key as keyof object]
      }
    }
  }

  public static withOnlyRequestedKeys = <T>(obj: object, keys: string[]): T => {
    const result: T = {} as T
    for (const key of keys) {
      if (Object.hasOwn(obj, key)) {
        result[key as keyof T] = obj[key as keyof object]
      }
    }

    return result
  }

  public static getUniqueSortedNonEmpty = (values: string[]): string[] => {
    return Array.from(new Set<string>(values.filter(v => !!v)).values()).sort()
  }

  public static isNumber = (s: unknown) => {
    return (!!s || s === 0) && !isNaN(Number(s))
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public static debounce<F extends (...args: any[]) => any>(fn: F, ms = 250) {
    let timer: ReturnType<typeof setTimeout> | null = null

    return function (...args: Parameters<F>): Promise<ReturnType<F>> {
      clearTimeout(timer as ReturnType<typeof setTimeout>)

      return new Promise(resolve => {
        timer = setTimeout(() => resolve(fn(...args)), ms)
      })
    }
  }

  public static downloadFileFromURL = (url: string) => {
    const anchor = document.createElement('a')
    anchor.setAttribute('target', '_blank')
    anchor.href = url
    anchor.style.display = 'none'
    // you can't click a link unless it's part of the document:
    document.body.appendChild(anchor)
    anchor.click()
    document.body.removeChild(anchor)
  }

  public static getUTCDateString = (date: Date): string => date.toLocaleDateString(undefined, {timeZone: 'UTC'})

  public static sleep = (ms: number) => new Promise(resolve => setTimeout(resolve, ms))

  public static getPercentage = (value: number): string => {
    return value > 0 && value < 1 ? '<1%' : `${Math.round(value)}%`
  }

  public static getDuration = (ms: number): string => {
    // number => xd xh xm xs

    if (ms < 1000) {
      return `<1${LABELS.time.seconds}`
    }

    let remainingTime = ms
    let days = Math.floor(remainingTime / TIME_IN_MS.day)

    remainingTime = remainingTime - days * TIME_IN_MS.day
    let hours = Math.floor(remainingTime / TIME_IN_MS.hour)

    remainingTime = remainingTime - hours * TIME_IN_MS.hour
    let minutes = Math.floor(remainingTime / TIME_IN_MS.minute)

    remainingTime = remainingTime - minutes * TIME_IN_MS.minute
    let seconds = Math.round(remainingTime / TIME_IN_MS.second)

    if (hours === 24) {
      days++
      hours = 0
    }
    if (minutes === 60) {
      hours++
      minutes = 0
    }
    if (seconds === 60) {
      minutes++
      seconds = 0
    }

    const daysStr = days > 0 ? `${days}${LABELS.time.days}` : ''
    const hoursStr = hours > 0 ? `${hours}${LABELS.time.hours}` : ''
    const minutesStr = minutes > 0 ? `${minutes}${LABELS.time.minutes}` : ''
    const secondsStr = seconds > 0 ? `${seconds}${LABELS.time.seconds}` : ''

    return [daysStr, hoursStr, minutesStr, secondsStr].filter(str => !!str).join(' ') || ''
  }

  public static getCardHeading = (
    title: string,
    description: string,
    dateRange: string,
    format?: (value: number) => string,
  ): CardHeading => {
    return {
      title,
      description: `${description} ${dateRange}`,
      format,
    }
  }

  public static columnValidForScope(id: string, scope: ScopeType) {
    if (scope === ScopeType.Org) {
      return true
    }

    if (scope === ScopeType.Repo) {
      return id !== COLUMNS.repository
    }
  }
}
