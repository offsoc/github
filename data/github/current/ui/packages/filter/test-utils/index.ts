import {act, fireEvent, screen} from '@testing-library/react'

export const setupExpectedAsyncErrorHandler = () => {
  jest.spyOn(console, 'error').mockImplementation((message: string) => {
    // * Because Filter is asynchronous, there are console errors that are thrown, but expected. This will rethrow
    // * any errors that are not related to the async nature of the component.
    if (!message.includes?.('wrapped in act(')) {
      // eslint-disable-next-line no-console
      console.warn(message)
    }
  })
}

export async function updateFilterValue(key: string, value: string = '') {
  jest.useFakeTimers()

  const input = screen.getByRole('combobox')
  input.focus()

  fireEvent.change(input, {target: {value: `${key}${key && value ? ':' : ''}${value}`}})

  // eslint-disable-next-line @typescript-eslint/require-await
  await act(async () => {
    jest.runAllTimers()
  })
}
