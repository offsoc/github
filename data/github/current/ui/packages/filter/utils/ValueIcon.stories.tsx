import {IssueOpenedIcon, OrganizationIcon} from '@primer/octicons-react'
import type {Meta, StoryObj} from '@storybook/react'

import {AvatarType} from '../types'
import {ValueIcon} from '.'

const meta = {
  title: 'Recipes/Filter/ValueIcon',
  component: ValueIcon,
  tags: ['!autodocs'],
} satisfies Meta<typeof ValueIcon>

export default meta

type Story = StoryObj<typeof ValueIcon>

export const GitHubAvatar: Story = {
  args: {
    value: {
      avatar: {
        type: AvatarType.User,
        url: 'https://avatars.githubusercontent.com/u/2447456?s=40&v=4',
      },
      value: '@dusave',
    },
  },
}
GitHubAvatar.storyName = 'GitHub Avatar'

export const Icon: Story = {
  args: {
    value: {
      icon: IssueOpenedIcon,
      iconColor: 'var(--fgColor-success, var(--color-success-fg))',
      value: 'Issue Open',
    },
  },
}

export const ProviderIcon: Story = {
  args: {
    value: {
      icon: OrganizationIcon,
      value: 'Organization',
    },
  },
}

export const ColorDot: Story = {
  args: {
    value: {
      iconColor: 'var(--fgColor-closed, var(--color-closed-fg))',
      value: 'Closed',
    },
  },
}
