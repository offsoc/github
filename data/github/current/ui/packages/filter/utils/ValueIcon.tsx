import {GitHubAvatar} from '@github-ui/github-avatar'
import type {Icon} from '@primer/octicons-react'
import {Box, Octicon} from '@primer/react'

import type {FilterValueData} from '../types'
import {getFilterValue} from '.'

export interface ValueIconProps {
  value: FilterValueData
  providerIcon?: Icon
  squareIcon?: boolean
}

export const ValueIcon = ({value, providerIcon, squareIcon = false}: ValueIconProps) => {
  if (value.avatar?.url) {
    return (
      <GitHubAvatar
        src={value.avatar.url}
        alt={getFilterValue(value.value) ?? 'User Avatar'}
        square={squareIcon}
        sx={{minWidth: '20px'}}
      />
    )
  }
  if (value.iconColor && !value.icon) {
    return <Box sx={{bg: value.iconColor, borderRadius: 9, width: '10px', height: '10px'}} />
  }
  if (value.icon) {
    const cssIconVariable = value.iconColor ? `${value.iconColor} !important` : 'currentcolor'
    return <Octicon icon={value.icon} sx={{fill: cssIconVariable}} />
  }
  if (providerIcon) {
    return <Octicon icon={providerIcon} />
  }

  return null
}
