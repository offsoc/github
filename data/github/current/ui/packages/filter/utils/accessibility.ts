import {announce} from '@github-ui/aria-live'

// This is a delay we configure to avoid screen readers from reading the suggestions list immediately on change
const SCREEN_READER_FEEDBACK_TIMEOUT_MS = 150

let screenReaderTimeout: NodeJS.Timeout | null

export const updateScreenReaderFeedback = (message: string = '') => {
  if (screenReaderTimeout) clearTimeout(screenReaderTimeout)
  screenReaderTimeout = setTimeout(() => {
    announce(message)
  }, SCREEN_READER_FEEDBACK_TIMEOUT_MS)
}
