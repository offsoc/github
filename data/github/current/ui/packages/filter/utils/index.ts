import {NoEntryIcon, PlusCircleIcon, SkipIcon} from '@primer/octicons-react'
import cloneDeep from 'lodash-es/cloneDeep'

import {FILTER_VALUES} from '../constants/filter-constants'
import type {FilterQuery} from '../filter-query'
import {RawTextProvider} from '../providers/raw'
import type {StateFilterProviderType} from '../providers/static'
import {
  type AnyBlock,
  type ARIAFilterSuggestion,
  BlockType,
  type ComplexFilterOperator,
  type FilterBlock,
  type FilterConfig,
  FilterOperator,
  type FilterProvider,
  FilterProviderType,
  type FilterSuggestion,
  type FilterSuggestionGroup,
  FilterTypeOperators,
  FilterValueType,
  type IndexedAnyBlock,
  type IndexedBlockKey,
  type IndexedBlockValueItem,
  type IndexedFilterBlock,
  type IndexedGroupBlock,
  type IndexedTextBlock,
  type IndexedUnmatchedCloseParenBlock,
  type IndexedUnmatchedOpenParenBlock,
  type MutableFilterBlock,
} from '../types'

export {ValueIcon} from './ValueIcon'

export const capitalize = (str: string) => str.charAt(0).toUpperCase() + str.slice(1)

export const escapeString = (str?: string) => {
  if (!str || (str.indexOf(' ') < 0 && str.indexOf('(') < 0 && str.indexOf(')') < 0)) return str

  let result = str
  if (!result.startsWith('"')) result = `"${result}`
  if (!result.endsWith('"')) result = `${result}"`

  return result
}

export const unescapeString = (str?: string) => {
  if (!str || str.replaceAll('"', '').length < 1) return str
  let result = str
  if (result.startsWith('"')) result = result.substring(1)
  if (result.endsWith('"')) result = result.substring(0, result.length - 1)

  return result
}

export const caseInsensitiveStringCompare = (str1?: string, str2?: string) => {
  return !!str1 && !!str2 && str1.toLocaleLowerCase() === str2.toLocaleLowerCase()
}

export const getFilterValueByCaretIndex = (values: IndexedBlockValueItem[], caretIndex: number | null = -1) => {
  if (caretIndex === null || caretIndex < 0) return getFilterValue(values.at(-1)?.value)
  for (const value of values) {
    if (isBetweenInclusive(caretIndex, value.startIndex, value.endIndex)) {
      return getFilterValue(value.value)
    }
  }
  return null
}

export const getProviderKey = (key: string) => {
  return key.startsWith('-') ? key.slice(1) : key
}

export const getUniqueReactKey = (context: string, id?: string, fallback?: string) => {
  return `${context}-${id ?? fallback}`
}

export const getFilterValue = (value: string | (() => string) | null | undefined) => {
  if (!value) return null
  if (typeof value === 'string') return value
  return value()
}

export const getUnescapedFilterValue = (value: string | (() => string) | null | undefined) => {
  const val = getFilterValue(value)
  return val ? unescapeString(val) : val
}

export const getEscapedFilterValue = (value: string | (() => string) | null | undefined) => {
  const val = getFilterValue(value)
  return val ? escapeString(val) : val
}

export const promiseTimeout = (ms: number) => {
  return new Promise(resolve => setTimeout(resolve, ms))
}

/*
  Compares whether the joined values of the filter query blocks matches the provided raw string or the filter query raw string as a fallback.
*/
export const checkFilterQuerySync = (filterQuery: FilterQuery, rawString?: string): boolean => {
  return (
    filterQuery.blocks
      .map(({raw}) => raw)
      .join('')
      .trim() === (rawString !== undefined ? rawString : filterQuery.raw).trim()
  )
}

export const isBetweenInclusive = (index: number, start: number, end: number): boolean => {
  return index >= 0 && index >= start && index <= end
}

export const isFilterBlock = (filterBlock: AnyBlock | MutableFilterBlock): filterBlock is FilterBlock => {
  return filterBlock.type === BlockType.Filter && filterBlock.key !== undefined && filterBlock.value !== undefined
}

export const isIndexedAnyBlock = (filterBlock: AnyBlock): filterBlock is IndexedAnyBlock => {
  return (
    (filterBlock as IndexedAnyBlock).startIndex !== undefined && (filterBlock as IndexedAnyBlock).endIndex !== undefined
  )
}

export const isIndexedGroupBlock = (filterBlock: AnyBlock): filterBlock is IndexedGroupBlock => {
  return (
    filterBlock.type === BlockType.Group &&
    (filterBlock as IndexedGroupBlock).blocks !== undefined &&
    (filterBlock as IndexedGroupBlock).startIndex !== undefined &&
    (filterBlock as IndexedGroupBlock).endIndex !== undefined
  )
}

export const isIndexedFilterBlock = (filterBlock: AnyBlock | MutableFilterBlock): filterBlock is IndexedFilterBlock => {
  return (
    filterBlock.type === BlockType.Filter &&
    (filterBlock as IndexedFilterBlock).key !== undefined &&
    (filterBlock as IndexedFilterBlock).value !== undefined &&
    (filterBlock as IndexedFilterBlock).startIndex !== undefined &&
    (filterBlock as IndexedFilterBlock).endIndex !== undefined
  )
}

export const isIndexedTextBlock = (textBlock: AnyBlock): textBlock is IndexedTextBlock => {
  return (
    textBlock.type === BlockType.Text &&
    (textBlock as IndexedTextBlock).startIndex !== undefined &&
    (textBlock as IndexedTextBlock).endIndex !== undefined
  )
}

export const isIndexedUnmatchedOpenParenBlock = (
  unmatchedOpenParenBlock: AnyBlock,
): unmatchedOpenParenBlock is IndexedUnmatchedOpenParenBlock => {
  return (
    unmatchedOpenParenBlock.type === BlockType.UnmatchedOpenParen &&
    (unmatchedOpenParenBlock as IndexedUnmatchedOpenParenBlock).startIndex !== undefined &&
    (unmatchedOpenParenBlock as IndexedUnmatchedOpenParenBlock).endIndex !== undefined
  )
}

export const isIndexedUnmatchedCloseParenBlock = (
  unmatchedCloseParenBlock: AnyBlock,
): unmatchedCloseParenBlock is IndexedUnmatchedCloseParenBlock => {
  return (
    unmatchedCloseParenBlock.type === BlockType.UnmatchedCloseParen &&
    (unmatchedCloseParenBlock as IndexedUnmatchedCloseParenBlock).startIndex !== undefined &&
    (unmatchedCloseParenBlock as IndexedUnmatchedCloseParenBlock).endIndex !== undefined
  )
}

export const isMutableFilterBlock = (filterBlock: AnyBlock | MutableFilterBlock): filterBlock is MutableFilterBlock => {
  return filterBlock.type === BlockType.Filter
}

export const isComplexFilterOperator = (
  operator: FilterOperator[] | ComplexFilterOperator,
): operator is ComplexFilterOperator => {
  return (operator as ComplexFilterOperator).single !== undefined
}

export const findExistingValueUsage = (filterQuery: FilterQuery, key: string, value: string | null | undefined) => {
  if (!value) return []
  return filterQuery.blocks.filter(
    f => isFilterBlock(f) && f.provider.key === key && f.value.values.findIndex(v => v.value === value) > -1,
  )
}

export const getAllFilterOperators = (provider?: FilterProvider): FilterOperator[] => {
  const operator = FilterTypeOperators[provider?.type ?? FilterProviderType.Unknown]
  const multiValue = provider?.options?.filterTypes.multiValue

  let operatorList = isComplexFilterOperator(operator) ? operator[multiValue ? 'multi' : 'single'] : operator

  if (provider?.options?.filterTypes.inclusive === false) {
    operatorList = operatorList.filter(o => o !== FilterOperator.Is && o !== FilterOperator.IsOneOf)
  }
  if (provider?.options?.filterTypes.exclusive === false) {
    operatorList = operatorList.filter(o => o !== FilterOperator.IsNot && o !== FilterOperator.IsNotOneOf)
  }

  return operatorList
}

export const getBlockKey = (provider: FilterProvider) =>
  provider.type === FilterProviderType.RawText ? undefined : {value: provider.key, valid: true}

export const getFilterOperator = (
  provider?: FilterProvider,
  filterKey?: string,
  filterValue?: string,
): FilterOperator => {
  const operators = getAllFilterOperators(provider)
  const multiValue = provider?.options?.filterTypes.multiValue
  if (operators[0]) {
    if (filterKey?.startsWith('-')) {
      return (
        operators.find(m => (multiValue ? m === FilterOperator.IsNotOneOf : m === FilterOperator.IsNot)) ?? operators[0]
      )
    }

    if (provider?.type === FilterProviderType.Number) {
      if (filterValue?.includes('..')) {
        return operators.find(m => m === FilterOperator.Between) ?? operators[0]
      } else if (filterValue?.startsWith('<=')) {
        return operators.find(m => m === FilterOperator.LessThanOrEqualTo) ?? operators[0]
      } else if (filterValue?.startsWith('<')) {
        return operators.find(m => m === FilterOperator.LessThan) ?? operators[0]
      } else if (filterValue?.startsWith('>=')) {
        return operators.find(m => m === FilterOperator.GreaterThanOrEqualTo) ?? operators[0]
      } else if (filterValue?.startsWith('>')) {
        return operators.find(m => m === FilterOperator.GreaterThan) ?? operators[0]
      }
    } else if (provider?.type === FilterProviderType.Date) {
      if (filterValue?.includes('..')) {
        return operators.find(m => m === FilterOperator.Between) ?? operators[0]
      } else if (filterValue?.startsWith('<=')) {
        return operators.find(m => m === FilterOperator.BeforeAndIncluding) ?? operators[0]
      } else if (filterValue?.startsWith('<')) {
        return operators.find(m => m === FilterOperator.Before) ?? operators[0]
      } else if (filterValue?.startsWith('>=')) {
        return operators.find(m => m === FilterOperator.AfterAndIncluding) ?? operators[0]
      } else if (filterValue?.startsWith('>')) {
        return operators.find(m => m === FilterOperator.After) ?? operators[0]
      }
    }
    return operators.find(m => (multiValue ? m === FilterOperator.IsOneOf : m === FilterOperator.Is)) ?? operators[0]
  }
  return FilterOperator.Is
}

export const buildRawBlockString = (filterBlock: Omit<MutableFilterBlock, 'raw'>, config: FilterConfig) => {
  const delimitedKey = filterBlock.key ? `${filterBlock.key.value}:` : ''

  switch (filterBlock.operator) {
    case FilterOperator.IsNot:
    case FilterOperator.IsNotOneOf:
      return `${delimitedKey.startsWith('-') ? delimitedKey : `-${delimitedKey}`}${filterBlock.value?.values
        .map(v => getEscapedFilterValue(v.value))
        .join(config.valueDelimiter)}`
    case FilterOperator.Before:
    case FilterOperator.LessThan:
      return `${delimitedKey}<${filterBlock.value?.raw ?? ''}`
    case FilterOperator.After:
    case FilterOperator.GreaterThan:
      return `${delimitedKey}>${!filterBlock.value?.raw ? '' : filterBlock.value?.raw}`
    case FilterOperator.LessThanOrEqualTo:
      return `${delimitedKey}<=${!filterBlock.value?.raw ? '' : filterBlock.value?.raw}`
    case FilterOperator.GreaterThanOrEqualTo:
      return `${delimitedKey}>=${!filterBlock.value?.raw ? '' : filterBlock.value?.raw}`
    case FilterOperator.Between:
      return `${delimitedKey}${getFilterValue(filterBlock.value?.values[0]?.value)}..${getFilterValue(
        filterBlock.value?.values[1]?.value,
      )}`

    default:
      return `${delimitedKey.startsWith('-') ? delimitedKey.substring(1) : delimitedKey}${filterBlock.value?.values
        .map(v => getEscapedFilterValue(v.value))
        .join(config.valueDelimiter)}`
  }
}

export const getNoValueSuggestion = (displayName: string, icon = NoEntryIcon, priority = 0): ARIAFilterSuggestion => {
  return {
    displayName: `No ${displayName.toLocaleLowerCase()}`,
    ariaLabel: `No ${displayName}, ${displayName}`,
    icon,
    value: '',
    type: FilterValueType.NoValue,
    priority,
  }
}

export const getExcludeKeySuggestion = (value?: string) => {
  return {
    displayName: value ? `Exclude ${value.toLocaleLowerCase()}` : 'Exclude',
    ariaLabel: value ? `Exclude ${value.toLocaleLowerCase()}` : 'Exclude',
    value: value ? `-${value}` : '-',
    type: FilterValueType.Key,
    priority: 1,
    icon: NoEntryIcon,
  }
}

export const getAndKeySuggestion = () => {
  return {
    ariaLabel: 'And',
    value: 'AND',
    type: 'keyword' as const,
    priority: 1,
    icon: PlusCircleIcon,
  }
}

export const getOrKeySuggestion = () => {
  return {
    ariaLabel: 'Or',
    value: 'OR',
    type: 'keyword' as const,
    priority: 1,
    icon: SkipIcon,
  }
}

export const getFlatSuggestionsList = (
  suggestions: ARIAFilterSuggestion[] | FilterSuggestionGroup[],
): ARIAFilterSuggestion[] => {
  const flatSuggestions: ARIAFilterSuggestion[] = []
  for (const suggestion of suggestions) {
    if ('suggestions' in suggestion) {
      flatSuggestions.push(...suggestion.suggestions)
    } else {
      flatSuggestions.push(suggestion)
    }
  }
  return flatSuggestions
}

export const getMutableFilterBlocks = (filteredBlocks: AnyBlock[]): MutableFilterBlock[] => {
  const blocks: MutableFilterBlock[] = []

  for (let i = 0; i < filteredBlocks.length; i++) {
    const block = filteredBlocks[i]
    if (!block) continue
    else if (block.type === BlockType.Space) continue
    else if (block.type === BlockType.Text && filteredBlocks[i - 2]?.type === BlockType.Text) {
      const prevBlock = blocks.pop()
      if (prevBlock) {
        prevBlock.raw += ` ${block.raw}`
        prevBlock.value = {raw: prevBlock.raw, values: [{value: prevBlock.raw, valid: true}]}
        blocks.push(prevBlock)
      }
    } else if (isMutableFilterBlock(block)) {
      const safeBlock: MutableFilterBlock = cloneDeep(block)

      // If the block value is between two values, we split them for visual display
      if (safeBlock.operator === FilterOperator.Between) {
        const joinedValue = getFilterValue(safeBlock.value?.values[0]?.value) ?? ''
        const values = joinedValue?.split('..')
        const validity = !!values[0] && values[0] !== '' && !!values[1] && values[1] !== ''
        if (values?.length === 2) {
          safeBlock.value = {
            raw: joinedValue,
            values: [
              {value: values[0] ?? '', valid: validity},
              {value: values[1] ?? '', valid: validity},
            ],
          }
        }
      } else {
        let raw = ''
        safeBlock.value?.values.map(v => {
          raw += `,${(getFilterValue(v.value) ?? '').replaceAll(/[>|<|>=|<=]/g, '') ?? ''}`
        })
        raw = raw.slice(1)
        safeBlock.value = {
          values: safeBlock.value?.values ?? [],
          raw,
        }
      }
      safeBlock.id = blocks.length
      blocks.push(safeBlock)
    } else {
      blocks.push({
        type: BlockType.Filter,
        id: blocks.length,
        provider: RawTextProvider,
        raw: block.raw ?? '',
        value: {raw: block.raw ?? '', values: [{value: block.raw ?? '', valid: true}]},
      })
    }
  }

  return blocks
}

export const getLastFilterBlockValue = (filterBlock: AnyBlock | MutableFilterBlock, caretIndex?: number | null) => {
  let lastValue: string
  if (isIndexedFilterBlock(filterBlock) && filterBlock.value?.values) {
    lastValue = getFilterValueByCaretIndex(filterBlock.value.values, caretIndex) ?? ''
  } else if (isFilterBlock(filterBlock) && filterBlock.value?.values) {
    lastValue = getFilterValue(filterBlock.value.values[0]?.value) ?? ''
  } else {
    lastValue = filterBlock.raw
  }
  if (lastValue.startsWith('"')) lastValue = lastValue.substring(1)
  return lastValue
}

export const isEmptyFilterBlockValue = (filterBlock: AnyBlock | MutableFilterBlock): boolean => {
  const filterBlockValues = isFilterBlock(filterBlock) && filterBlock.value?.values ? filterBlock.value.values : []

  if (filterBlockValues.length > 1) return false

  return filterBlockValues.length === 1 ? filterBlockValues[0]?.value === '' : true
}

export const hasFilterBlockFocusChanged = (filterBlock: AnyBlock, newCaretIndex: number) => {
  if (!isIndexedFilterBlock(filterBlock)) return true

  if (!filterBlock.key.hasCaret) {
    return isBetweenInclusive(newCaretIndex, filterBlock.key.startIndex, filterBlock.key.endIndex)
  }
  const currentValue = filterBlock.value.values.find(v => v.hasCaret)
  return currentValue && !isBetweenInclusive(newCaretIndex, currentValue.startIndex, currentValue.endIndex)
}

export const shouldProviderShowSuggestions = (filterBlock: AnyBlock | MutableFilterBlock, isMultiValue = true) =>
  isFilterBlock(filterBlock) && (isMultiValue || filterBlock.value.values.length <= 1)

export const getFilterBlockChunkByCaret = (
  filterBlock: IndexedFilterBlock,
  caretIndex: number,
): [type: 'key', IndexedBlockKey] | [type: 'value', IndexedBlockValueItem] | [null, null] => {
  // If the caret is before or after the block, return null. #notmyblock
  if (filterBlock.startIndex > caretIndex || filterBlock.endIndex < caretIndex) return [null, null]

  // If the caret is in the key, return the key
  if (isBetweenInclusive(caretIndex, filterBlock.key.startIndex, filterBlock.key.endIndex))
    return ['key', filterBlock.key]

  const currentValue = filterBlock.value.values.find(v => isBetweenInclusive(caretIndex, v.startIndex, v.endIndex))
  return currentValue ? ['value', currentValue] : [null, null]
}

export const getFilterValuesByStateType = (type: StateFilterProviderType = 'mixed') => {
  let values: FilterSuggestion[] = []
  if (type === 'mixed') {
    const uniqueValues: Record<string, FilterSuggestion> = {}
    for (const value of [...FILTER_VALUES.memexState, ...FILTER_VALUES.state, ...FILTER_VALUES.prState]) {
      const extractedValue = getFilterValue(value.value)

      const {icon, iconColor, ...keepAttrs} = value
      if (extractedValue && !uniqueValues[extractedValue]) {
        uniqueValues[extractedValue] = {...keepAttrs}
      }
    }
    values = Object.values(uniqueValues)
  } else if (type === 'issues') {
    values = FILTER_VALUES.state
  } else if (type === 'memex') {
    values = FILTER_VALUES.memexState
  } else {
    values = FILTER_VALUES.prState
  }

  return values
}

export function sanitizeOperators(raw: string | null): string | null {
  return raw ? raw.replaceAll('<', '&lt;')?.replaceAll('>', '&gt;') : raw
}
