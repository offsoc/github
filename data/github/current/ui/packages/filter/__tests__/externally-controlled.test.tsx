import {render} from '@github-ui/react-core/test-utils'
import {screen} from '@testing-library/react'

import {updateFilterValue} from '../test-utils'
import {ExternalWrapper} from './utils/ExternalWrapper'
import {
  expectFilterValueToBe,
  expectSuggestionsToBeEmpty,
  expectSuggestionsToMatchSnapshot,
  setupAsyncErrorHandler,
} from './utils/helpers'

describe('Externally Controlled Filter', () => {
  setupAsyncErrorHandler()

  it('should not show the suggestions menu if the input is not focused', async () => {
    const {user} = render(<ExternalWrapper />)

    await updateFilterValue('state', 'o')

    expectSuggestionsToMatchSnapshot()

    await user.click(document.body)

    const revertButton = await screen.findByTestId('filter-revert-query')
    await user.click(revertButton)

    await expectSuggestionsToBeEmpty()

    expectFilterValueToBe('')
  })
})
