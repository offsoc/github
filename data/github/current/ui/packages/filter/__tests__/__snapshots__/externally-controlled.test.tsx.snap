// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Externally Controlled Filter should not show the suggestions menu if the input is not focused 1`] = `
[
  "Open",
  "Closed",
]
`;
